base/multidimensional.jl

```diff
@@ -334,20 +340,26 @@ module IteratorsMD
         iterfirst, iterfirst
     end
     @inline function iterate(iter::CartesianIndices, state)
-        nextstate = CartesianIndex(inc(state.I, first(iter).I, last(iter).I))
-        nextstate.I[end] > last(iter.indices[end]) && return nothing
-        nextstate, nextstate
+        return inc((), state.I, first(iter).I, last(iter).I)
     end
 
     # increment & carry
-    @inline inc(::Tuple{}, ::Tuple{}, ::Tuple{}) = ()
-    @inline inc(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int}) = (state[1]+1,)
-    @inline function inc(state, start, stop)
+    # increment post check to avoid integer overflow
+    @inline inc(out, ::Tuple{}, ::Tuple{}, ::Tuple{}) = nothing
+    @inline function inc(out, state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int})
```

Comment:
Do we even need this method anymore? I think it's covered by the generic case.

Reply:
I don't think so. I added it because this is a common pattern elsewhere (probably steming from a time when inference liked code like this more).